File: chat-backend/src/main/java/com/smallchat/backend/chat/application/usecase/SaveMessageUseCase.java
package com.smallchat.backend.chat.application.usecase;

import org.springframework.stereotype.Service;

import com.smallchat.backend.chat.application.inputport.SaveMessageInputPort;
import com.smallchat.backend.chat.application.outputport.MessageOutputPort;
import com.smallchat.backend.chat.domain.event.MessagePublished;
import com.smallchat.backend.chat.domain.model.vo.Message;
import com.smallchat.backend.chat.domain.model.vo.Sender;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class SaveMessageUseCase implements SaveMessageInputPort {

    private final MessageOutputPort messageOutputPort;

    @Override
    public void saveMessage(MessagePublished messagePublished) {
        String nickname = messagePublished.getNickname();
        String userId = messagePublished.getUserId();
        String message = messagePublished.getMessage();
        String chatId = messagePublished.getChatId();
        messageOutputPort.save(new Message(message, chatId, new Sender(userId, nickname)));
    }
}

File: chat-backend/src/main/java/com/smallchat/backend/chat/application/usecase/SaveMessageUseCase.kt
package com.smallchat.backend.chat.application.usecase

import com.smallchat.backend.chat.domain.interfaces.MessageSimpleQueryRepository
import com.smallchat.backend.chat.domain.model.MessageKt
import org.springframework.stereotype.Service

@Service
class SaveMessageUseCase(private val messageSimpleQueryRepository: MessageSimpleQueryRepository) {
    fun execute(messageKt: MessageKt) {
        val (chatId, userId, nickname, message) = messageKt
        messageSimpleQueryRepository.save(
            MessageKt.userMessage(
                chatId,
                userId,
                nickname,
                message,
            )
        )
    }
}
